File: revalidate.R
# place this file in C:\Program Files\R\"whatever R version is installed"\bin>
# in the command prompt, go and run in C:\Program Files\R\"whatever R version is installed"\bin>r -f revalidate.R

# iteratively execute curl commands to revalidate and update records in CDH
# iteration frequency is based on records retrieved from elasticsearch

# download packages and establish system connection
require(elastic)
require(httr)
require(dplyr)
connect()

# read in all zips from depots
setwd("C:/Users/Administrator/Documents")
file <- read.csv("Depot_PLZ_Zuordnung_2016_12.csv", sep=";", colClasses = c("depot"="character", "postcode"="character"))
total_count <- 0
zip_collector <- list()

# query all zips in file and count affected records
for(zip in file[,1]){
  frag1_q <- '{ "query": { "match": { "zip": "'
  frag2_q <- '"}}}'
  match <- paste0(frag1_q, as.character(zip), frag2_q, "")
  # choose a valid index name
  res <- Search(index="customer-d01", body=match, scroll="5m", search_type = "scan")
  total_count <- total_count + res$hits$total
  zip_collector <- append(zip_collector, zip)
  if(total_count > 350000){
    break
  }
}
# write zips to file to be picked up by admin command
print("Requested amount of zips has been fetched. Writing to file path now. Starting revalidation process.")
Sys.sleep(2)
df_zip_collector <- t(as.data.frame(zip_collector))
out_file <- paste("C:\\uniserv\\cdh\\temp\\plzdelta\\delta.csv", sep="")
write.csv2(df_zip_collector, out_file, row.names = FALSE, quote = FALSE)
# adapt origin file so it will pick up correctly when program is scheduled 
row_number_current_zip <- which(file$postcode == zip)
outf.df <- slice(file, row_number_current_zip:length(file[,1]))
setwd("C:/Users/Administrator/Documents")
out_file <- paste(getwd(), "/Depot_PLZ_Zuordnung_2016_12.csv", sep="")
write.csv2(outf.df, out_file, row.names = FALSE, quote = FALSE)
# system command
shell(paste("cd",setwd("C:/uniserv/cdh/tools"),sep=" "))
Sys.sleep(2)
# enter valid curl command
shell(paste('curl.exe', ' -X', ' POST ', 'http://localhost:6441/current/database/xxx/_revalidate-by-zip?SingleCsv=C:\\uniserv\\cdh\\temp\\plzdelta\\delta.csv', ' --data', ' " "', sep=""))
print("Revalidation of zip code has finished. Now starting to update historical names.")
Sys.sleep(2)
shell(paste("cd",setwd("C:/uniserv/cdh/tools"),sep=" "))
# call update of historical names based on processed zips
for(zip in zip_collector){
  print(paste0("updating historical data to zip code ", as.character(zip), sep = ""))
  shell(paste('curl.exe', ' -X', ' POST ', 'http://localhost:6441/current/database/dpd/_histname-update-batch?ZipCode=', as.character(zip), ' --data', ' " "', sep=""))
}
print("That's all for today folks!")

File: revalidate.py
# import relevant libraries

import os
import csv
import pandas
import elasticsearch
from elasticsearch import Elasticsearch
es = Elasticsearch()

# read in all zip files from depots
os.chdir('C:\\Users\\Administrator\\Documents')
file = pandas.read_csv('Depot_PLZ_Zuordnung_2016_12.csv', sep = ';', converters={'depot': str, 'postcode': str})
frag1 = '{"query": { "match": { "zip":"'
frag2 = '"}}}'

total_count = []
zipCollector = []

# query all zips in file and count records respectively
for zip in file['postcode']:
    zipCollector.append(zip)
    match = frag1 + zip + frag2
    # choose valid index name
    res = es.count(index="customer-d01", body=match)
    total_count.append(res['count'])
    if total_count > 400000:
       break

# write collected zips to delta.csv to be picked up by cdh
print("Requested amount of zips has been fetched. Writing to file path now. Starting revalidation process.")
zipCollector = pandas.DataFrame(zipCollector)
pandas.DataFrame.to_csv(zipCollector, path_or_buf="C:\\uniserv\\cdh\\temp\\plzdelta\\delta.csv", index = False, encoding="UTF-8", header=False)
# slice data to contain everything from last cutoff and write to wd
crtrow = file[file['postcode'] == zip].index.tolist()
out_df = file[(crtrow[0]+1):len(file)]
path = os.getcwd() + "\\Depot_PLZ_Zuordnung_2016_12.csv"
pandas.DataFrame.to_csv(out_df, path_or_buf=path, index = False, encoding="UTF-8", header=False, sep = ";")
# invoke shell commands to call revalidation and update-histnames
os.chdir("C:\\Uniserv\\cdh\\tools")
os.system("admin help")
# choose valid command
os.system('curl.exe' + ' -X' + ' POST ' + 'http://localhost:6441/current/database/xxx/_revalidate-by-zip?SingleCsv=C:\\uniserv\\cdh\\temp\\plzdelta\\delta.csv' + ' --data' + ' " "')
print("Revalidation of zip code has finished. Starting to update historical names.")
# choose valid command
os.system('curl.exe' + ' -X' + ' POST ' + 'http://localhost:6441/current/database/xxx/_revalidate-by-zip?SingleCsv=C:\\uniserv\\cdh\\temp\\plzdelta\\delta.csv' + ' --data' + ' " "')
for zip in zipCollector:
    os.system('curl.exe' + ' -X' + ' POST ' + 'http://localhost:6441/current/database/dpd/_histname-update-batch?ZipCode=' + zip + ' --data' + ' " "')
print("End of processing")




